notebook/nbextensions.py

```diff
@@ -299,118 +355,300 @@ def _config_file_name_default(self):
-        sys.exit("Please supply at least one subcommand: %s" % subcmds)
+        self.log.warn("Please supply at least one subcommand: %s" % subcmds)
+        sys.exit(1)
 
 main = NBExtensionApp.launch_instance
 
+#------------------------------------------------------------------------------
+# Private API
+#------------------------------------------------------------------------------
+
+
+def _should_copy(src, dest, verbose=1, log=None):
+    """Should a file be copied?"""
+    if not os.path.exists(dest):
+        return True
+    if os.stat(src).st_mtime - os.stat(dest).st_mtime > 1e-6:
+        # we add a fudge factor to work around a bug in python 2.x
+        # that was fixed in python 3.x: http://bugs.python.org/issue12904
+        if verbose >= 2:
+            log("%s is out of date" % dest)
+        return True
+    if verbose >= 2:
+        log("%s is up to date" % dest)
+    return False
+
+
+def _maybe_copy(src, dest, verbose=1, log=None):
+    """Copy a file if it needs updating."""
+    if log is None: log = print
+    if _should_copy(src, dest, verbose=verbose, log=log):
+        if verbose >= 1:
+            log("Copying: %s -> %s" % (src, dest))
+        shutil.copy2(src, dest)
+
+
+def _safe_is_tarfile(path):
+    """Safe version of is_tarfile, return False on IOError."""
+    try:
+        return tarfile.is_tarfile(path)
+    except IOError:
+        return False
+
+
+def _get_nbextension_dir(user=False, sys_prefix=False, prefix=None, nbextensions_dir=None):
+    """Return the nbextension directory specified"""
+    if sum(map(bool, [user, prefix, nbextensions_dir, sys_prefix])) > 1:
+        raise ArgumentConflict("cannot specify more than one of user, sys_prefix, prefix, or nbextensions_dir")
+    if user:
+        nbext = pjoin(jupyter_data_dir(), u'nbextensions')
+    elif sys_prefix:
+        nbext = pjoin(ENV_JUPYTER_PATH[0], u'nbextensions')
+    elif prefix:
+        nbext = pjoin(prefix, 'share', 'jupyter', 'nbextensions')
+    elif nbextensions_dir:
+        nbext = nbextensions_dir
+    else:
+        nbext = pjoin(SYSTEM_JUPYTER_PATH[0], 'nbextensions')
+    return nbext
+
+
+def _get_config_dir(user=False, sys_prefix=False):
+    if sum(map(bool, [user, sys_prefix])) > 1:
+        raise ArgumentConflict("Cannot specify more than one of user or sys_prefix")
+    if user:
+        nbext = jupyter_config_dir()
+    elif sys_prefix:
+        nbext = ENV_CONFIG_PATH[0]
+    else:
+        nbext = SYSTEM_CONFIG_PATH[0]
+    return nbext
+
+
+def _get_nbextension_metadata(package):
+    m = __import__(package)
+    if not hasattr(m, '_jupyter_nbextension_paths'):
+        raise KeyError('The Python package {} is not a valid nbextension'.format(package))
+    nbexts = m._jupyter_nbextension_paths()
+    return m, nbexts
+
+
+def _read_config_data(user=False, sys_prefix=False):
+    config_dir = _get_config_dir(user=user, sys_prefix=sys_prefix)
+    config_file = os.path.join(config_dir, 'jupyter_notebook_config.json')
```

Comment:
it would be nice if `jupyter_notebook_config.json` was exposed as a constant someplace: but hopefully nobody should ever have to deal with this again!